ENH: hard-code finfo parameters for known types

matthew-brett opened the pull request:
Hard-code the MachAr parameters for float64, float32, 80-bit extended
precision, to save time, and provide skeleton for more difficult types
such as the double double on PPC; see
https://github.com/numpy/numpy/issues/2669

matthew-brett committed:
ENH: hard-code finfo parameters for known types

Hard-code the MachAr parameters for float64, float32, 80-bit extended
precision, to save time, and provide skeleton for more difficult types
such as the double double on PPC; see
https://github.com/numpy/numpy/issues/2669

matthew-brett commented:
There's a test error from `test_warnings.py` in https://travis-ci.org/numpy/numpy/jobs/193586749#L3330 . It's telling me I'm not allowed to use `warnings.simplefilter('ignore')` - which I used [here](https://github.com/numpy/numpy/pull/8504/files#diff-898819c3f8146d8ec2980500ffb602f3R35).  It seems like a perfectly sensible use to me - why is `test_warnings` complaining?

charris commented:
Hmm, not sure. @seberg Thoughts?

charris commented:
```
    Context manager and decorator doing much the same as
    ``warnings.catch_warnings``.

    However, it also provides a filter mechanism to work around
    http://bugs.python.org/issue4180.

    This bug causes Python before 3.4 to not reliably show warnings again
    after they have been ignored once (even within catch_warnings). It
    means that no "ignore" filter can be used easily, since following
    tests might need to see the warning. Additionally it allows easier
    specificity for testing warnings and can be nested.
```
So the `"ignore"` is the problem. Note also that the warning filters are not thread safe, so should be avoided outside of tests when possible.

matthew-brett commented:
Sure - but how should I do what I'm trying to do if I'm not allowed to ignore the warning inside the context manager block?

matthew-brett commented:
OK - found a `suppress_warnings` incantation that seems to work.

njsmith commented:
What's the warning here?

matthew-brett committed:
ENH: add PPC long double finfo

Add parameters for PPC long double, fixing long-standing bug for finfo
on PPC.

seberg commented:
Well, to be honest, for some code doing funny warning stuff may make sense and the test should likely be relaxed to only tests. On the other hand, I think `np.seterr` is probably the correct thing here, since it seems to me that this should be floating point stuff (and I think seterr is even threadsafe? frankly not sure).

seberg commented:
Or if the warning makes sense/does not matter, moving the context to the tests makes more sense.

pv commented:
seterr is threadsafe iirc

njsmith commented:
`seterr` is threadsafe yeah: https://github.com/numpy/numpy/blob/20d6ca1bd0c23cf22880ea72ae7abb4b67d41894/numpy/core/src/umath/ufunc_object.c#L4517

mraspaud commented:
@matthew-brett on ppc64, I get a mantissa (significand) size of 116 bits for long doubles (double double in this case). But you seem to set it to 105, (`105, 1024, 16`), may I ask why ?

matthew-brett commented:
OK - I used the `errstate` context manager instead, that does look better.

matthew-brett commented:
@mraspaud - I got the number of digits from looking at the `eps` value found by using `np.nextafter(np.longdouble(1), np.longdouble(2))`.  Actually, the number also makes some sense, if we suppose that the number of digits comes from the sum of the number of significand digits in both doubles = 53 + 53 - 1 for the implicit first.

charris commented:
I was thinking of something more along the lines of the build time determination, basically viewing a value as a string and then looking that up. See `numpy/core/setup_common.py`.

matthew-brett commented:
Chuck - 

a) I think we'd need most of this machinery anyway, to replicate the old finfo behavior.  Then the only difference between build-time and run-time detection is the small fragment of code here to detect the `eps` and `huge` and byte size values, and select accordingly;
b) I can't explain the mechanism, but I believe I have seen instances where the type of longdouble changed at run time - see comment [here](https://github.com/nipy/nibabel/blob/master/nibabel/casting.py#L684).

matthew-brett committed:
BF: use Python floats for float64 finfo

For some reason (garrgh) np.exp2 with float64 gives a different answer on
Windows than it does on other platforms; use Python floating point
calculations instead, which do appear to be consistent.

matthew-brett commented:
Other somewhat trivial advantage of this approach - finfo is a lot faster.  With this PR:

```
In [2]: time np.finfo(np.float64)
CPU times: user 189 µs, sys: 23 µs, total: 212 µs
Wall time: 218 µs
```

With 1.12 release wheel:

```
In [2]: time np.finfo(np.float64)
CPU times: user 12.1 ms, sys: 308 µs, total: 12.4 ms
Wall time: 12.3 ms
```

Of course, this is a one-time only cost, as the results are cached.

matthew-brett commented:
@mraspaud - also see https://en.wikipedia.org/wiki/Quadruple-precision_floating-point_format#Double-double_arithmetic - which says these have 106 digits in the significand.   The numpy figures for significand digits are all N-1, I guess omitting the implicit first.  Also see : http://mrob.com/pub/math/f161.html (the paper explains that using the sign bit of the second double makes the actual number 107...).

matthew-brett commented:
Longdoubles are especially slow in the old implementation:

```
In [3]: time np.finfo(np.longdouble)
CPU times: user 150 ms, sys: 2.02 ms, total: 152 ms
Wall time: 161 ms
Out[3]: finfo(resolution=1e-18, min=-1.18973149536e+4932, max=1.18973149536e+4932, dtype=float128)
```

Same command takes 263 microseconds from code in this PR.

mraspaud commented:
@matthew-brett indeed, I was just mislead by HDF5 reporting 116 bits significand...

matthew-brett commented:
Chuck - would you consider merging this one, with a hoped-for follow-up to do build-time detection?

charris commented:
@matthew-brett I'll take a look at this later today. I wasn't actually suggesting doing build-time detection, but rather doing the same sort of type detection at runtime. It is/can be all done in python. But I will need to take a closer look here before deciding if it is worthwhile.

matthew-brett commented:
OK - thanks.  If you're unhappy with the float32, float64, float80 detection, I'm happy to drop those, leaving only the PPC longdouble, which does need a fix.

charris commented on numpy/core/getlimits.py:
```diff
@@ -9,19 +9,198 @@
+    key = (digits, maxexp, bytes)
+    if key in _KNOWN_TYPES:
+        return _KNOWN_TYPES[key]
+    return _discover_machar(dtype)
```
Maybe just raise an error? We only guarantee IEEE, and now the double double. There is also a Motorola extended precision, I don't know if we should worry about that.

charris replied:
Or warning?

charris commented on numpy/core/getlimits.py:
```diff
@@ -9,19 +9,198 @@
+# Key to identify the floating point type is tuple (number of bits in
+# significand (ignoring implicit first digit), maximum exponent, number of
+# bytes).
+_KNOWN_TYPES = {(52, 1024, 8): _float64_ma,
```
Here I'd like to eventually get rid of nextafter and just have a big dictionary keyed by byte strings, where the string is generated by something like
```
>>> np.array(1234, dtype='g').tostring()
b'\x00\x00\x00\x00\x00\x00@\x9a\t@\xd73-V\x00\x00'
```
for all the types. I don't know how stable/unique the encoding is, however, and we would need both little and big endian versions. Signed/unsigned characters could be a problem perhaps.

charris replied:
Make that `tobytes` for clarity.

charris replied:
Lookup is pretty fast
```
In [11]: d =  {array(-123456789.0, 'g').tobytes(): 'long double'}

In [12]: timeit d[array(-123456789.0, 'g').tobytes()]
1000000 loops, best of 3: 506 ns per loop
```

charris commented:
OK, I think this works. Long term I'd like to simplify, tabulate more of the info, and get rid of `MachAr`. I don't know how important backward compatibility would be, depends on how many folks directly access the results of `MachAr`.

charris commented:
@matthew-brett This could use a release note as it should improve behavior of long double in PPC.

mraspaud commented:
@matthew-brett Just a note:
according to this document:
https://www.ibm.com/support/knowledgecenter/en/ssw_aix_71/com.ibm.aix.genprogc/128bit_long_double_floating-point_datatype.htm
, it looks like the mantissa size is not absolutely fixed in the type definition:
> Because 128-bit numbers usually provide at least 106 bits of precision

However it looks like in GCC the mantissa is fixed to 106 bits.
https://gcc.gnu.org/wiki/Ieee128PowerPC

matthew-brett committed:
DOC: add release notes for finfo fixes

Add release note describing fixes for PPC long double ``finfo``.

matthew-brett commented:
@mraspaud - thanks for the links.  I've used one of them in the release note entry.

I think the problem with the digits in the significand is the one pointed out in the IBM entry; the number is composed of two floating point values where the overall value is the sum of the two parts:  F = H + L where H is the high-order float64, and L is the low-order float64.

Quoting the IBM article:

> If the low-order part is much less than 1 ULP of the high-order part, significant bits (either all 0's or all 1's) are implied between the significant of the high-order and low-order numbers. 

For example, imagine the floats were base-10 floats with 2 digits in the significand. Superficially, this gives us 4 digits in the significand, in total, but:

H = 1.0e00
L = 1.0e-10
F = 1.0000000001

charris commented on numpy/core/getlimits.py:
```diff
@@ -9,19 +9,198 @@
+    start = ftype(1)
+    direction = ftype(direction)
+    above = start + direction
+    return (nextafter(start, above) - start) * direction
```
For negative direction this is going to depend on rounding of the difference. If you actually used that option it would bother me more ;)

matthew-brett replied:
Could you explain more?  Won't `nextafter(1, -1) - 1)` always be exactly representable?  

```
In [13]: for dt in np.sctypes['float']:
    ...:     assert _get_eps(dt, -1) == _discover_machar(dt).epsneg
```

charris replied:
I don't know that it is ever a problem, I think float arithmetic is done with two quard bits for rounding, but if you shift `nextafter(1, 0)` to align it with 1.0, the last bit goes off the end.

charris commented:
@matthew-brett The IBM doc says that `nextafterl` is not available on AIX and it looks like we don't have a fallback for that.

charris commented:
Also, 16 bit floats are IEEE, so might want to hard code those also.

charris commented:
What system have you tested the double double finfo on? I'm a bit curious as the "true" nextafter value is much smaller that `1 + eps`, so if this works it sounds like the library definition of nextafter has been adapted.

matthew-brett commented:
How can I detect when `nextafter` is available?   `hasattr(numpy, 'nextafter')` ?

matthew-brett commented:
I tested on PPC 64 (Mac G4, big-endian).   Yes, it gives this result above, yes, you're right, that's not actually the next representable number, I guess it must be the library.  I suppose that could depend on the compiler ...

matthew-brett commented:
For my reference - perl configure code to detect long double format - search for 'double double' in https://perl5.git.perl.org/perl.git/blob/3118d7d684b56cbeb702af874f4326683c45f045:/Configure

charris commented:
I think we can check for signatures, which we can get from the setup. For instance on my intel
```
In [18]: a = array(0, 'g')

In [19]: b = zeros(a.itemsize, uint8)

In [20]: b.view('g')[...]=-123456789.0

In [21]: b
Out[21]: 
array([  0,   0,   0,   0, 160, 162, 121, 235,  25, 192,   0,   0,   0,
         0,   0,   0], dtype=uint8)
```
The zeroing is needed because what goes in the `16 - 10 == 6` left over bytes looks to be random garbage otherwise.

matthew-brett commented:
How about something like:

```
In [28]: ftype = np.float64

In [29]: np.array(-2**10-1, np.dtype(ftype).newbyteorder('<')).tobytes()
Out[29]: b'\x00\x00\x00\x00\x00\x04\x90\xc0'
```

?

charris commented:
The problem is garbage bytes in the extended precision versions. IEEE80 only needs 8 bytes, but is stored in 12 or 16. What shows up in the byte string doesn't seem consistent. However, once it is zeroed we should be OK as long as the byte string is always uses unsigned characters. Need to check that...

charris commented:
OK, `tobytes` calls `PyBytes_FromStringAndSize` which just copies the bytes over to a byte string. So I guess the question if whether the displayed string is always the same, since that is what we want to use for the key. I think it is always ascii with other byte values done in `\x**` form, so probably OK.

charris commented:
Looks like the extended_double_double form reverses the two doubles separately to go from big to litttle endian and back. You should check that that works correctly with `newbyteorder`.

matthew-brett committed:
RF: try using byte string to identify floats

From suggestion by Chuck, and
https//perl5.git.perl.org/perl.git/blob/3118d7d684b56cbeb702af874f4326683c45f045:/Configure

matthew-brett commented:
Code works correctly on PPC 64 big-endian (G4 machine).  I don't have anything else to test on.  Does anyone have access to a ppc64le machine?

mraspaud commented:
@matthew-brett I do, what do you want me to test ?

matthew-brett commented:
@mraspaud - can you post the result of running this:

```
import numpy as np
print(repr(np.longdouble('-0.1').newbyteorder('<').tobytes()))
```

mraspaud commented:
```
(h5py)debian@debian8-ppc64el:~$ python
Python 2.7.9 (default, Aug 13 2016, 16:27:51) 
[GCC 4.9.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
>>> print(repr(np.longdouble('-0.1').newbyteorder('<').tobytes()))
'\x9a\x99\x99\x99\x99\x99\xb9\xbf\x9a\x99\x99\x99\x99\x99Y<'
>>> 
(h5py)debian@debian8-ppc64el:~$ uname -a
Linux debian8-ppc64el 3.16.0-4-powerpc64le #1 SMP Debian 3.16.36-1+deb8u2 (2016-10-19) ppc64le GNU/Linux
```

matthew-brett committed:
TST: add tests for reasonable finfo parameters

Check that finfo returns somewhat reasonable parameters for all floating
point types.

matthew-brett commented:
@mraspaud - can you test the results of:

```
nosetests path/to/numpy/core/tests/test_getlimits.py
```

with the current branch?

matthew-brett commented:
I think this one is ready now.   The detection doesn't cover IEEE float128, hence the lack of warning when the byte-string detection fails and falls back to the discovery code.  If anyone has an AIX machine they have access to, it would be good to check on that - but at least, this code falls back to the previous code when it fails to detect a signature, so it shouldn't be worse at detection.

matthew-brett commented:
Planning on getting access to a SPARC with actual IEEE 754 128-bit floats to test against.

mraspaud commented:
```
(h5py)debian@debian8-ppc64el:~$ nosetests numpy/numpy/core/tests/test_getlimits.py 
.............
----------------------------------------------------------------------
Ran 13 tests in 0.062s

OK
```

charris commented on numpy/core/getlimits.py:
```diff
@@ -8,20 +8,239 @@
+        if key in _KNOWN_TYPES:
+            return _KNOWN_TYPES[key]
+    # Fall back to parameter discovery
+    return _discovered_machar(ftype)
```
Might be worth raising a warning here so that we might get reports of new types we should cover.

charris replied:
Maybe include the missing key in the message.

matthew-brett committed:
RF: add warning for not-recognized float type

Warn if we don't have a signature for the float type.

matthew-brett committed:
NF: add IEEE float128 type signature

Still needs test on platform with IEEE float128, to check MachAr-like
parameters are correct.

pv commented:
nextafter is AFAIK present in umath on all platforms and defined for all 
float types. I think things are written so that Numpy compilation will 
fail if the long double float format is not recognized (cf 
src/private/npy_fpmath.h, src/npymath/npy_math_private.h).
.
Can the numpy C code that understands the float formats be leveraged 
here, instead of needing to reimplement?

charris commented:
@pv The C code just writes a structure that is then read and analysed in python, but is otherwise similar to the current code. Nextafter isn't really defined for double_double, so depends on the library choices, although the libraries are likely to follow the IBM recommendation. If we ever expose the build time macros the code here can be modified, but as it stands should serve as good check for hardcoding the types. The only thing that worries me a bit is that the `-0.1` value is not exactly representable, so the result may depend on the library implementation. OTOH, it is a test of the libraries.

charris commented:
I should say that by C code analysed is the object file and the Python code looks for the structure in that file.

matthew-brett commented:
Chuck - I believe (though happy to be corrected) that all the floating point implementations specify a unique correspondence between bit pattern and input number (here -0.1).  I mean, they guarantee the closest representable number, and there is only one bit pattern for any given representable number.  I can see how there might be more than one pattern for double double (1.0 + 0.01 == 1.1 - 0.09), but I haven't looked into it.  The patterns we're getting also match with those recorded in the perl configure script I pointed to - [here](https://perl5.git.perl.org/perl.git/blob/3118d7d684b56cbeb702af874f4326683c45f045:/Configure) - so I guess we can depend on them, but in any case, all that will happen, if they are wrong, is that the code will return to the original behavior.

matthew-brett commented:
Tests pass on SPARC, with IEEE float128.  Ready to merge from my point of view.

charris commented:
In it goes, thanks @matthew-brett .

charris commented:
How do the timings compare?

matthew-brett commented:
Timings, macOS Intel, numpy 1.12:

```
In [3]: time np.finfo(np.longdouble)
CPU times: user 139 ms, sys: 2.43 ms, total: 141 ms
Wall time: 149 ms
```

This branch:

```
In [2]: time np.finfo(np.longdouble)
CPU times: user 165 µs, sys: 23 µs, total: 188 µs
Wall time: 195 µs
```

ahaldane commented:
@matthew-brett This commit is causing me a bit of trouble as described in [this comment](https://github.com/numpy/numpy/pull/8983#discussion_r115838683).

The problem revolves around the fact that you are calling numpy functions before numpy modules are fully loaded here, specifically you are calling `array2string`. This causes cyclic import issues.

Do you think we might delay the evaluations here until after numpy modules are fully loaded?

matthew-brett commented:
@ahaldane - does this work?  https://github.com/numpy/numpy/pull/9113
